Add Schechter1D model

larrybradley opened the pull request:
### Description

This PR adds a `Schechter1D` model for a Schechter luminosity function parameterized in terms of magnitudes.


<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->


<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


### Checklist for package maintainer(s)
<!-- This section is to be filled by package maintainer(s) who will
review this pull request. -->

This checklist is meant to remind the package maintainer(s) who will review this pull request of some common things to look for. This list is not exhaustive.

- [x] Do the proposed changes actually accomplish desired goals?
- [x] Do the proposed changes follow the [Astropy coding guidelines](https://docs.astropy.org/en/latest/development/codeguide.html)?
- [x] Are tests added/updated as required? If so, do they follow the [Astropy testing guidelines](https://docs.astropy.org/en/latest/development/testguide.html)?
- [x] Are docs added/updated as required? If so, do they follow the [Astropy documentation guidelines](https://docs.astropy.org/en/latest/development/docguide.html#astropy-documentation-rules-and-guidelines)?
- [x] Is rebase and/or squash necessary? If so, please provide the author with appropriate instructions. Also see ["When to rebase and squash commits"](https://docs.astropy.org/en/latest/development/when_to_rebase.html).
- [x] Did the CI pass? If no, are the failures related? If you need to run daily and weekly cron jobs as part of the PR, please apply the `Extra CI` label.
- [x] Is a change log needed? If yes, did the change log check pass? If no, add the `no-changelog-entry-needed` label. If this is a manual backport, use the `skip-changelog-checks` label unless special changelog handling is necessary.
- [x] Is this a big PR that makes a "What's new?" entry worthwhile and if so, is (1) a "what's new" entry included in this PR and (2) the "whatsnew-needed" label applied?
- [x] Is a milestone set? Milestone must be set but `astropy-bot` check might be missing; do not let the green checkmark fool you.
- [x] At the time of adding the milestone, if the milestone set requires a backport to release branch(es), apply the appropriate `backport-X.Y.x` label(s) *before* merge.

pllim commented:
Probably needs what's new?

pllim commented:
Also add user facing doc at https://docs.astropy.org/en/latest/modeling/index.html#pre-defined-models ?

larrybradley commented:
I added a "what's new" entry.

pllim commented on astropy/modeling/powerlaws.py:
```diff
@@ -507,3 +507,115 @@ def input_units(self):
+        import numpy as np
+
+        phi_star = 4.3e-4 * (u.Mpc ** -3)
+        m_star = -20.26
```
Does this model support the magnitude units?

larrybradley replied:
No.  `10 ** <mag Quantity>` fails (gives an error), so `m_star` cannot have mag units.  I state this in the docs.

larrybradley commented:
@pllim I added a new page for the powerlaw models (it was completely missing).  I also fixed the formatting of the polynomial models so the 1D and 2D models are properly indented under polynomial models.

larrybradley commented:
The codestyle test was failing because of a newly-implemented `isort` test.  `isort` apparently didn't like that my max line length was 80 char instead of 100, which is completely braindead.  It insists on make the lines *longer*.  Why are we doing this???

WilliamJamieson commented on astropy/modeling/powerlaws.py:
```diff
@@ -507,3 +507,115 @@ def input_units(self):
+    @staticmethod
+    def evaluate(mag, phi_star, m_star, alpha):
+        """Schechter luminosity function model function."""
+        factor = 10 ** (0.4 * (m_star - mag))
```
@larrybradley you mentioned that the reason why `m_star` cannot have units is this due to this computation.

I suggest you break this up into something like:
```python
factor_exp = 0.4 * (m_star - mag)

if isinstance(factor_exp, Quantity):
    factor_exp = factor_exp.value

factor = 10 ** factor_exp
```
This should fix the units issue.

larrybradley replied:
Thanks, @WilliamJamieson.  I added this, including checks for magnitude units (to prevent, e.g., `m_star = 10 * u.km`).

pllim replied:
I am a little worried... what if someone defines `m_star` in ABmag and then pass in `mag` in STmag? Is that possible?

larrybradley replied:
I assumed that would raise an error, but in fact it does not!  That is not good.  So I need to add checks that `m_star` and `mag` have the same (mag) unit.

larrybradley replied:
I added a check that `m_star.unit == mag.unit`.

pllim commented:
@nstarman or @WilliamJamieson , is there a way to make isort more flexible?

WilliamJamieson commented:
> @nstarman or @WilliamJamieson , is there a way to make isort more flexible?

I don't think so, the easy way to avoid the code-style issue in general is just to install pre-commit: https://pre-commit.com/ and then enable it to run as a git pre commit hook. Most of the failures will be caught and/or fixed when running `git commit`.

larrybradley commented:
> @nstarman or @WilliamJamieson , is there a way to make isort more flexible?
> 
> I don't think so, the easy way to avoid the code-style issue in general is just to install pre-commit: https://pre-commit.com/ and then enable it to run as a git pre commit hook. Most of the failures will be caught and/or fixed when running git commit.

This is very unfriendly to new contributors to the project, most of whom are not software developers.

WilliamJamieson commented:
I should clarify my previous statement about pre-commit, one can also just run `isort` on the file(s) changed if one does not want to set up pre-commit.

> This is very unfriendly to new contributors to the project, most of whom are not software developers.

It is a trade off... `isort` is part of an effort to improve the general code "quality" and consistency that is present. While it does impose a little more work on contributors, it does provide automatic tools (running `isort` command or pre-commit) which will adjust your contributions to fit the "standards" without you having to manually edit the code to meet them.

WilliamJamieson reviewed (changes requested):
Now that units are better supported can you add this model to the `test_models_quantities.py` test suite as well?

larrybradley commented:
> I should clarify my previous statement about pre-commit, one can also just run isort on the file(s) changed if one does not want to set up pre-commit.
> 
> This is very unfriendly to new contributors to the project, most of whom are not software developers.
> 
> It is a trade off... isort is part of an effort to improve the general code "quality" and consistency that is present. While it does impose a little more work on contributors, it does provide automatic tools (running isort command or pre-commit) which will adjust your contributions to fit the "standards" without you having to manually edit the code to meet them.

As a very long-time contributor I'm sensitive to the experience of new contributors (we want to increase the number of new contributors).   I'm all for code standards and consistency, but there must be a balance.  IMHO, `isort` failing on perfectly acceptable (and "standard") code because it wants to *increase* my line length is completely brain dead.  That kind of stuff scares off new contributors.

pllim commented:
Does `test_models_quantities.py` even test against the use case where user try to mix up different mag systems? If not, I think we need to write a new test also to prevent regression.

larrybradley commented:
> Now that units are better supported can you add this model to the test_models_quantities.py test suite as well?

I've added the model here and the tests pass (locally), *BUT* only if the unit is `u.mag`.  `m_star` and `mag` in `u.MagUnit` units (e.g., u.ABmag or u.STmag) fails.  I'm not sure that models are actually allowed to have `MagUnits`.  Simple example:

```
from astropy.modeling.models import Schechter1D
import astropy.units as u

unit = u.ABmag
phi_star = 1.e-4 * (u.Mpc ** -3)
m_star = -20. * unit
alpha = -1.9
model = Schechter1D(phi_star, m_star, alpha)
model(-23 * unit)
```
Gives:
`UnitTypeError: Quantity instances require normal units, not <class 'astropy.units.function.logarithmic.MagUnit'> instances.`

But if you change `unit = u.mag`, it works fine.  `u.mag` is a `astropy.units.function.mixin.RegularFunctionUnit` not a `MagUnit`.

Perhaps it's best after all to not allow mag units.

pllim commented:
There is conflict in the what's new entry, please rebase.

@WilliamJamieson , is this PR good to go? If so, please approve.

Also, let's discuss the `isort` stuff in a separate issue. It is out of scope here. Thanks!

larrybradley commented:
Rebased!

pllim commented:
Thanks, all!
